### ISSUE TITLE
The app doesn't compare between the local version and the published one

### ISSUE THREAD

**Issue by MarcoIeni** [obi1kenobi/cargo-semver-checks#309]
### Steps to reproduce the bug with the above code

```sh
git clone https://github.com/rust-github/rust-gh-example.git
cd rust-gh-example
cargo semver-checks check-release
```

output:

```
$ cargo semver-checks check-release
    Updating index
     Parsing rust-gh-example v0.1.8 (current)
     Parsing rust-gh-example v0.1.7 (baseline)
    Checking rust-gh-example v0.1.7 -> v0.1.7 (no change)
   Completed [   0.008s] 33 checks; 33 passed, 0 unnecessary
```

Question: why does it compare 0.1.7 with 0.1.7? I would expect it compares current and baseline.

Anyway, now I try to break the api by adding the `#[must_use]` attribute:

```rust
#[must_use]
pub fn answer() -> u32 {
    42
}
``` 

I run `cargo semver-checks check-release` again.

### Actual Behaviour

The check still pass:

```
$ cargo semver-checks check-release
    Updating index
     Parsing rust-gh-example v0.1.8 (current)
     Parsing rust-gh-example v0.1.7 (baseline)
    Checking rust-gh-example v0.1.7 -> v0.1.7 (no change)
   Completed [   0.008s] 33 checks; 33 passed, 0 unnecessary
```

### Expected Behaviour

I would expect the program to tell me I have added the must_use attribute.

#### Software version

cargo-semver-checks 0.15.2

#### Operating system

macOS 13.1 (Darwin 22.2.0)

#### Command-line

```bash
/Users/marcoieni/.cargo/bin/cargo-semver-checks semver-checks --bugreport
```

#### cargo version

```
> cargo -V
cargo 1.66.1 (ad779e08b 2023-01-10)
```

#### Compile time information

- Profile: release
- Target triple: aarch64-apple-darwin
- Family: unix
- OS: macos
- Architecture: aarch64
- Pointer width: 64
- Endian: little
- CPU features: aes,crc,dit,dotprod,dpb,dpb2,fcma,fhm,flagm,fp16,frintts,jsconv,lor,lse,neon,paca,pacg,pan,pmuv3,ras,rcpc,rcpc2,rdm,sb,sha2,sha3,ssbs,vh
- Host: aarch64-apple-darwin

### Additional Context

I'm trying to use this library, but I didn't find a way to simulate a semver violation. I'm sure I'm missing something, but I tried many things. How can I test this library locally?
I'm confused by the log line `Checking rust-gh-example v0.1.7 -> v0.1.7 (no change)`. Why doesn't it compare with the version I have locally (0.1.8)?

**Comment by MarcoIeni** [comment]
It's a problem with my machine. I am with a friend with a linux machine and it works for him 👀

**Comment by tonowak** [comment]
Hi, thanks for the bug report!
I've tried to reproduce your issue with `v0.1.7 -> v0.1.7` with the steps you've provided and the same tool/Rust versions, but on my machine it correctly shows `v0.1.7 -> v0.1.8`.

I'll look into it a bit more soon, maybe I'll try it from a macos. Are you sure you don't have any custom build configurations? Can you try cloning the repo in a directory in `/tmp` and checking whether you have the same output from there?

For now I'll just say that it is probably only an issue with printing and the tool should work nevertheless (quite recently there have been a similar bug https://github.com/obi1kenobi/cargo-semver-checks/pull/248). So I think that the tool correctly compares the two crates.

> Anyway, now I try to break the api by adding the #[must_use] attribute:

I think this particular lint is not yet implemented (more details in `README.md` in section `Will cargo-semver-checks catch every semver violation?`). You can check the implemented lints through `cargo semver-checks check-release --list`. To test whether the tool works, you can e.g. add a new parameter to the function `answer` and it should report this change as a major change.

**Comment by MarcoIeni** [comment]
I have tried to break the api in different ways, such as renaming the function or adding parameters.
On linux the bug doesn't happen. On my m1 mac, it happens.

**Comment by tonowak** [comment]
Interesting, thanks for checking.

I propose to try the following steps -- they might help in finding the bug or help in reproducing the bug:
- create a directory outside of the home directory (in linux, the `/tmp` directory is a good choice), clone the repo with the code somewhere there and try to run the tool from it (the plan is to check whether the bug happens on a freshly cloned, non-touched repo, in a directory with a path that doesn't contain any weird characters and is outside of any local files in the vicinity that might affect the build process),
- if the tool still won't work, a good idea might be checking whether the issue is still present when passing a local baseline crate. You can do that by cloning the repo with the code to a different directory (so, let's say then you'll have a `rust-gh-example-current` with your modifications to the code, and `rust-gh-example-baseline` without any modifications to the code) and run the tool from the `rust-gh-example-current` directory as follows: `cargo semver-checks check-release --baseline-root=../rust-gh-example-baseline/Cargo.toml`

Additionally, could you provide any information about your local machine that might help me while trying to reproduce the bug? (I don't use a mac, so I'm not sure what kind of information might be beneficial for me to know; my plan is to setup a virtual machine with the same operating system as you).

**Comment by obi1kenobi** [comment]
Hey @MarcoIeni, thanks for reporting the bug and sorry you ran into it. Based on the symptoms, it sounds very likely to be another way to trigger #270. Just like @tonowak, I'm also very curious to figure out how it's happening.

TL;DR on that issue: `cargo-semver-checks` compares two rustdoc JSON files against each other. In some build configurations, rustdoc ends up overwriting the same file twice instead of generating two separate files. This causes the version number being checked to be wrong (it's loaded from the file) and also causes no semver issues to be found (because comparing the file against itself means nothing changed). That's 2/2 on obscure symptoms matching, which why @tonowak and I both think it's related.

@tonowak, thanks for the debugging help! It sounds like you've ruled out repo-local build configuration. Global build configuration in `~/.cargo/config.toml` is still a plausible explanation, though, and I think the steps you suggested won't rule it out. Let's do it directly, before you set up QEMU to emulate an m1 Mac :)

@MarcoIeni, can you try these for me first, in case we get lucky?
- Can you try running `cat ~/.cargo/config.toml` to see if there's any global (non-project-related) build configuration? If so, what is it?
- If you have no build configuration, then the build directory should be called `target` and exist side-by-side with the top-level `Cargo.toml` of the project. Is it indeed there?
- Try deleting the `target` directory and all its contents. Then, run `cargo semver-checks check-release` again. Which JSON files got created in the `target` directory and its subdirectories? What are their names and paths?

**Comment by tonowak** [comment]
Ooh, I see. So hopefully merging https://github.com/obi1kenobi/cargo-semver-checks/pull/270 will solve the issue. I'll try (or @staniewzki) to finish this PR today :)

**Comment by MarcoIeni** [comment]
This fixed it: `$ unset CARGO_TARGET_DIR`

**Comment by MarcoIeni** [comment]
I confirm that #270 fixes my issue! :)

**Comment by obi1kenobi** [comment]
Thanks! @tonowak and @staniewzki are on point for getting the PR polished up and merged, and I filed a PR and an issue to make the bug-reporting experience smoother based on what we learned here. Apologies for the trouble!
